src/ui/features/referral-program/ReferralProgramService.background.ts

```diff
@@ -0,0 +1,157 @@
+        messageContextParams: {
+          initiator: INTERNAL_ORIGIN,
+          clientScope: null,
+        },
+      },
+      context: { origin: INTERNAL_ORIGIN },
+    });
+
+    return signature;
+  }
+
+  async signAndSendReferWalletMessage({
+    address,
+    referralCode,
+  }: {
+    address: string;
+    referralCode: string;
+  }) {
+    const signature = await this.signReferralMessage({
+      address,
+      referralCode,
+    });
+    return ZerionAPI.referWallet({
+      address,
+      referralCode,
+      signature,
+    });
+  }
+
+  async applyReferralCodeToAllWallets({
+    referralCode,
+  }: {
+    referralCode: string;
+  }) {
+    invariant(this.options, "Options aren't expected to become null");
+    const walletFacade = this.options.getWallet();
+
+    const checkReferralResponse = await ZerionAPI.checkReferral({
+      referralCode,
+    });
+    const checkedReferrer = checkReferralResponse.data;
+    const walletGroups = await walletFacade.uiGetWalletGroups({
+      context: INTERNAL_SYMBOL_CONTEXT,
+    });
+
+    const signerWalletGroups = walletGroups?.filter((group) =>
+      isSignerContainer(group.walletContainer)
+    );
+    const ownedBareWallets =
+      signerWalletGroups?.flatMap(
+        (group) =>
+          group.walletContainer.wallets.filter(isBareWallet) as BareWallet[]
+      ) || [];
+    const ownedAddresses = ownedBareWallets.map((wallet) => wallet.address);
+
+    const walletsMetaResponse =
+      ownedAddresses.length > 0
+        ? await ZerionAPI.getWalletsMeta({ identifiers: ownedAddresses })
+        : null;
+    const walletsMeta = walletsMetaResponse?.data || [];
+
+    const eligibleAddresses = walletsMeta
+      .filter(
+        (meta) =>
+          meta.membership.referrer === null &&
+          meta.membership.referralCode !== referralCode
+      )
+      .map((meta) => meta.address);
+
+    await Promise.allSettled(
+      eligibleAddresses.map((address) =>
+        this.signAndSendReferWalletMessage({ address, referralCode })
+      )
+    );
+
+    await saveReferrerData(checkedReferrer);
+
+    // We need to start watching the referrer wallet
+    if (checkedReferrer.address) {
+      await walletFacade.uiAddReadonlyAddress({
+        context: INTERNAL_SYMBOL_CONTEXT,
+        params: {
+          address: checkedReferrer.address,
+          name: checkedReferrer.handle,
+        },
+      });
+    }
+
+    return checkedReferrer;
+  }
+
+  async applyReferralCodeToWallet({
+    walletContainer,
+  }: {
+    walletContainer: WalletContainer;
+  }) {
+    const currentReferrer = await readSavedReferrerData();
+    const referralCode = currentReferrer?.referralCode;
+
+    if (!isSignerContainer(walletContainer) || !referralCode) {
+      return;
+    }
+
+    await Promise.allSettled(
+      walletContainer.wallets.map((wallet) =>
+        this.signAndSendReferWalletMessage({
+          address: wallet.address,
+          referralCode,
+        })
+      )
+    );
```

Comment:
What happens to failed requests?

Reply:
For now, they’re simply ignored… Actually, I’m not even sure if we should try to handle failed requests at all

Reply:
Added error handling ✅